scripts/library/compression/calldataUtils.ts

```diff
@@ -0,0 +1,389 @@
+
+  const parts: Uint8Array[] = [];
+  for (let i = 0; i < chunks.length; i++) parts.push(toBeArray(chunks[i].data));
+  return parts;
+}
+
+// Chunking utils
+function bytesToChunks(data: string, length?: number): Chunk[] {
+  if (length === undefined) length = data.length / 2;
+  const fnsel = data.substring(0, 8);
+  const fndata = data.substring(8);
+  const expectedIndexEnd = (length - 4) * 2;
+  const padding = length == data.length / 2 ? undefined : ''.padEnd(data.length - length * 2, '0');
+
+  const params = SIGNATURE_MAP.get(fnsel);
+  if (params !== undefined) {
+    try {
+      const ch = chunk(fndata, 0, params);
+      if (ch.index != expectedIndexEnd) throw new Error('Incomplete chunking');
+
+      const chunks: Chunk[] = [{ type: 'fnsel', data: fnsel }, ...ch.chunks];
+      if (padding !== undefined) chunks.push({ type: 'padding', data: padding });
+      return chunks;
+    } catch (e) {}
+  }
+  return [{ type: 'bytes', data, length }];
+}
+function chunk(
+  data: string,
+  index: number,
+  params: ParamType[],
+  alreadyPointed: boolean = false,
+): { chunks: Chunk[]; index: number } {
+  const startIndex = index;
+  const chunks: Chunk[] = [];
+
+  //first pass through all params
+  const pointed: number[] = [];
+  for (let i = 0; i < params.length; i++) {
+    const param = params[i];
+    if (fixed(param) || alreadyPointed) {
+      let ch: { chunks: Chunk[]; index: number } = { chunks: [], index };
+      if (param.type == 'tuple') ch = chunk(data, index, param.params);
+      else if (param.type == 'array') ch = chunkArray(data, index, param.param);
+      else if (param.type == 'string') ch = chunkString(data, index);
+      else if (param.type == 'bytes') ch = chunkBytes(data, index);
+      else ch = chunkFixed(data, index, param);
+      chunks.push(...ch.chunks);
+      index = ch.index;
+    } else {
+      const err = 'Invalid formatting';
+      chunks.push({ type: 'pointer', data: safeSubstring(data, index, index + 64, err) });
+      pointed[i] = parseInt(safeSubstring(data, index, index + 64, err), 16) * 2;
+      index += 64;
+    }
+  }
+
+  //follow up on pointed data
+  for (let i = 0; i < pointed.length; i++) {
+    if (pointed[i]) {
+      const param = params[i];
+      let ch: { chunks: Chunk[]; index: number } = { chunks: [], index };
+      if (param.type == 'tuple') ch = chunk(data, startIndex + pointed[i], param.params);
+      else if (param.type == 'array') ch = chunkArray(data, startIndex + pointed[i], param.param);
+      else if (param.type == 'string') ch = chunkString(data, startIndex + pointed[i]);
+      else if (param.type == 'bytes') ch = chunkBytes(data, startIndex + pointed[i]);
+      chunks.push(...ch.chunks);
+      index = ch.index;
+    }
+  }
+
+  return { chunks, index };
+}
+function chunkArray(
+  data: string,
+  index: number,
+  param: ParamType,
+  length?: number,
+): { chunks: Chunk[]; index: number } {
+  //TODO: use length
```

Comment:
Is this a TODO for later?

Reply:
Fixed. The fast compression algorithm is still experimental and works on our test cases, but was designed to support many more untested cases.